codersdk/workspacesdk/connector_internal_test.go

```diff
@@ -142,6 +146,217 @@ func TestTailnetAPIConnector_UplevelVersion(t *testing.T) {
+	coordPtr.Store(&coord)
+	derpMapCh := make(chan *tailcfg.DERPMap)
+	defer close(derpMapCh)
+
+	clock := quartz.NewMock(t)
+	resumeTokenSigningKey, err := tailnet.GenerateResumeTokenSigningKey()
+	require.NoError(t, err)
+	resumeTokenProvider := tailnet.NewResumeTokenKeyProvider(resumeTokenSigningKey, clock, time.Hour)
+	svc, err := tailnet.NewClientService(tailnet.ClientServiceOptions{
+		Logger:                  logger,
+		CoordPtr:                &coordPtr,
+		DERPMapUpdateFrequency:  time.Millisecond,
+		DERPMapFn:               func() *tailcfg.DERPMap { return <-derpMapCh },
+		NetworkTelemetryHandler: func(batch []*proto.TelemetryEvent) {},
+		ResumeTokenProvider:     resumeTokenProvider,
+	})
+	require.NoError(t, err)
+
+	var (
+		websocketConnCh = make(chan *websocket.Conn, 64)
+		didFail         int64
+	)
+	svr := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if r.URL.Query().Get("resume_token") != "" {
+			atomic.AddInt64(&didFail, 1)
+			httpapi.Write(ctx, w, http.StatusUnauthorized, codersdk.Response{
+				Message: CoordinateAPIInvalidResumeToken,
+				Validations: []codersdk.ValidationError{
+					{Field: "resume_token", Detail: CoordinateAPIInvalidResumeToken},
+				},
+			})
+			return
+		}
+
+		sws, err := websocket.Accept(w, r, nil)
+		if !assert.NoError(t, err) {
+			return
+		}
+		testutil.RequireSendCtx(ctx, t, websocketConnCh, sws)
+		ctx, nc := codersdk.WebsocketNetConn(r.Context(), sws, websocket.MessageBinary)
+		err = svc.ServeConnV2(ctx, nc, tailnet.StreamID{
+			Name: "client",
+			ID:   uuid.New(),
+			Auth: tailnet.ClientCoordinateeAuth{AgentID: agentID},
+		})
+		assert.NoError(t, err)
+	}))
+
+	fConn := newFakeTailnetConn()
+
+	newTickerTrap := clock.Trap().NewTicker("tailnetAPIConnector", "refreshToken")
+	tickerResetTrap := clock.Trap().TickerReset("tailnetAPIConnector", "refreshToken", "reset")
+	defer newTickerTrap.Close()
+	uut := newTailnetAPIConnector(ctx, logger, agentID, svr.URL, clock, &websocket.DialOptions{})
+	uut.runConnector(fConn)
+
+	// Wait for the resume token to be fetched for the first time.
+	trappedTicker := newTickerTrap.MustWait(ctx)
+	trappedTicker.Release()
+	trappedReset := tickerResetTrap.MustWait(ctx)
+	trappedReset.Release()
+	originalResumeToken := uut.resumeToken.Token
+
+	// Sever the connection and expect it to reconnect with the resume token,
+	// which should fail and cause the client to be disconnected. The client
+	// should then reconnect with no resume token.
+	wsConn := testutil.RequireRecvCtx(ctx, t, websocketConnCh)
+	_ = wsConn.Close(websocket.StatusGoingAway, "test")
+
+	// Wait for the resume token to be refreshed, which indicates a successful
+	// reconnect.
+	trappedTicker = newTickerTrap.MustWait(ctx)
+	trappedTicker.Release()
+	// Since we failed the initial reconnect and we're definitely reconnected
+	// now, the stored resume token should now be nil.
+	require.Nil(t, uut.resumeToken)
```

Comment:
This is a racy assertion when placed here: you've released the call to `NewTicker` but have not yet trapped the call to `Reset`, which is exactly the time when `uut.resumeToken` gets updated.

If you move this assertion to just before `trappedTicker.Release()`, then `uut` is still trapped in the call to `NewTicker`, so must have already cleared the token and can't have refreshed it yet.